src/plugins/es_ui_shared/static/forms/hook_form_lib/hooks/use_field.ts

```diff
@@ -0,0 +1,360 @@
+
+    if (isEmptyString) {
+      return inputValue;
+    }
+    return formatters.reduce((output, formatter) => formatter(output), inputValue);
+  };
+
+  const onValueChange = () => {
+    setIsChangingValue(true);
+
+    // Update the form data observable
+    form.__updateFormDataAt(path, serializeOutput(value));
+
+    // Validate field(s) and set form.isValid flag
+    form.__validateFields(fieldsToValidateOnChange);
+
+    if (debounceTimeout.current) {
+      clearTimeout(debounceTimeout.current);
+    }
+
+    debounceTimeout.current = setTimeout(() => {
+      setIsChangingValue(false);
+    }, errorDisplayDelay);
+  };
+
+  const runValidations = ({
+    formData,
+    value: valueToValidate,
+    validationTypeToValidate,
+  }: {
+    formData: any;
+    value: unknown;
+    validationTypeToValidate?: string;
+  }): ValidationError[] | Promise<ValidationError[]> => {
+    let validationErrors: ValidationError[] = [];
+    let hasToRunAsync = false;
+
+    // By default, for fields that have an asynchronous validation
+    // we will clear the errors as soon as the field value changes.
+    clearErrors([VALIDATION_TYPES.FIELD, VALIDATION_TYPES.ASYNC]);
+
+    const runAsync = async () => {
+      for (const validation of validations) {
+        const {
+          validator,
+          exitOnFail = true,
+          type: validationType = VALIDATION_TYPES.FIELD,
+        } = validation;
+
+        if (
+          typeof validationTypeToValidate !== 'undefined' &&
+          validationType !== validationTypeToValidate
+        ) {
+          continue;
+        }
+
+        const validationResult = await validator({
+          value: (valueToValidate as unknown) as string,
+          errors: validationErrors,
+          form,
+          formData,
+          path,
+        });
+
+        if (!validationResult) {
+          continue;
+        }
+
+        validationErrors.push({
+          ...validationResult,
+          validationType: validationType || VALIDATION_TYPES.FIELD,
+        });
+
+        if (exitOnFail) {
+          break;
+        }
+      }
+
+      return validationErrors;
+    };
+
+    const runSync = () => {
+      // Sequencially execute all the validations for the field
+      for (const validation of validations) {
+        const {
+          validator,
+          exitOnFail = true,
+          type: validationType = VALIDATION_TYPES.FIELD,
+        } = validation;
+
+        if (
+          typeof validationTypeToValidate !== 'undefined' &&
+          validationType !== validationTypeToValidate
+        ) {
+          continue;
+        }
+
+        const validationResult: ValidationError = validator({
+          value: (valueToValidate as unknown) as string,
+          errors: validationErrors,
+          form,
+          formData,
+          path,
+        }) as ValidationError;
+
+        if (!validationResult) {
+          continue;
+        }
+
+        if (!!validationResult.then) {
```

Comment:
If I'm following this logic correctly, this means we first call `validator` and if it returns a promise, then we call `runAsync` which will call `validator` a second time. If the validator creates a network request, does that mean the request will be sent twice?

Reply:
Great point. I added support to cancel the validation. The consumer can optionally add a `cancel` method to the promise he returns (in case of an HTTP request the best would be to use axios that has a [cancelation API](https://github.com/axios/axios#cancellation). 

This `cancel()` method will be called before validating again the field.

Here is a "manual" example for a simple Promise:

```js
let counter = 0;

....

validations: [
  {
    validator: () => {
      let canceled = false;
      const current = counter;
      console.log('Validating', current);

      const promise = new Promise(resolve => {
        setTimeout(resolve, 2000);
      }).then(() => {
        if (canceled) {
          console.log('Has been canceled, should not do anything....', current);
        } else {
          console.log('Finished long validation!', current);
        }
      });
      // Attach a cancel() method
      (promise as any).cancel = () => (canceled = true);

      counter++;

      return promise;
    },
  },
],
```

And this is what is logged when entering "abc" in the field

```sh
Validating 0
Validating 1
Validating 2
Validating 3
Validating 4
Validating 5
Has been canceled, should not do anything.... 0
Has been canceled, should not do anything.... 1
Has been canceled, should not do anything.... 2
Has been canceled, should not do anything.... 3
Has been canceled, should not do anything.... 4
Finished long validation! 5
```